### PR TITLE
Draft: Score anchor flow

### PR DESCRIPTION
**Pull request by belyrs**
This PR is a draft and is created to gather reviews for the score anchor flow.

Let's consider that "demo/utils/updateScore" as a marketplace. Within this marketplace, the "fromJournalContent" function in "Scoring.ts" is invoked. The result produced by this function will be employed by the marketplace to trigger the "makeScoreEntryToChain" function in "Scoring.chain.ts."

### CODE REVIEW

File: packages/modules/src/scoring/Scoring.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  const ratingInput = fromScore(journalContent, creator)
+  const entry = ratingInput.entry
+  const scoreIdentifier = getUriForScore(entry)
+  return {
+    ratingInput,
+    scoreIdentifier,
+  }
+}
+
+export function verifyScoreStructure(input: IJournalContent) {
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts:91]: We have to expand this

File: packages/modules/src/scoring/Scoring.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  if (!input.collector) {
+    throw new SDKErrors.ScoreCollectorMissingError()
+  }
+  if (!input.entity) {
+    throw new SDKErrors.ScoreEntityMissingError()
+  }
+  if (!input.tid) {
+    throw new SDKErrors.ScoreTidMissingError()
+  }
+  if (!input.entry_type) {
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts]: check for approved entry_type

File: packages/modules/src/scoring/Scoring.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    throw new SDKErrors.ScoreEntryTypeMissingError()
+  }
+  if (!input.count) {
+    throw new SDKErrors.ScoreCountMissingError()
+  }
+  if (!input.rating) {
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts]: Should not be more that the max value. If, per rating entry max value is five, this entry should be within the limit of count * max_value

File: packages/modules/src/scoring/Scoring.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    throw new SDKErrors.ScoreRatingMissingError()
+  }
+  if (!input.rating_type) {
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts]: need to check the types

File: packages/types/src/Score.ts [hunk 5]
```diff
@@ -1,22 +1,33 @@
 import { HexString } from '@polkadot/util/types.js'
-import type { IPublicIdentity } from './PublicIdentity.js'
 
 export const SCORE_IDENTIFIER: number = 101
 export const SCORE_PREFIX: string = 'score:cord:'
+export const SCORE_MODULUS: number = 10
+export const SCORE_IDENT: number = 11034
 
 export enum ScoreType {
```
Comment (smohan-dw) [inline packages/types/src/Score.ts]: RatingType

File: packages/types/src/Score.ts [hunk 6]
```diff
@@ -1,22 +1,33 @@
...
-  delivery = 'delivery',
+  overall = 'Overall',
+  delivery = 'Delivery',
+}
+export enum EntryType {
+  credit = 'Credit',
+  debit = 'Debit',
 }
 
 export interface IJournalContent {
```
Comment (smohan-dw) [inline packages/types/src/Score.ts:18]: RatingEntry

File: packages/types/src/Score.ts [hunk 7]
```diff
@@ -40,11 +51,14 @@ export interface IScoreAggregateDetails {
     score: number
   }
 }
+
 export interface IScoreAverageDetails {
```
Comment (smohan-dw) [inline packages/types/src/Score.ts]: We don't need this at the SDK level. This computation should be done at the API level

File: packages/types/src/Score.ts [hunk 8]
```diff
@@ -40,11 +51,14 @@ export interface IScoreAggregateDetails {
...
-  average: {
-    count: number
-    score: number
-  }
+  rating: number
+  count: number
+  average: number
+}
+
+export interface IScoreDetails {
```
Comment (smohan-dw) [inline packages/types/src/Score.ts]: IEntityScoreDetails

File: packages/modules/src/scoring/Scoring.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    decodedScoreEntry.count
+  )
+  return {
+    rating: actualRating,
+    count: decodedScoreEntry.count,
+    average: averageRating,
+  }
+}
+
+export function fromJournalContent(
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts]: fromRatingEntry

File: packages/modules/src/scoring/Scoring.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  const hexDigest = Crypto.u8aToHex(digest)
+  return hexDigest
+}
+
+export function getUriForScore(journalContent: IJournalContent) {
+  const scoreDigest = generateDigestFromJournalContent(journalContent)
+  return Identifier.hashToUri(scoreDigest, SCORE_IDENT, SCORE_PREFIX)
+}
+
+export function fromScore(
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts]: get a better name for this function. This is transforming the ratingEntry. fromScore name will confuse devs

File: packages/modules/src/scoring/Scoring.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+  journalContent: IJournalContent,
+  creator: CordAddress
+): IRatingInput {
+  journalContent.rating = adjustAndRoundRating(journalContent.rating)
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.ts]: adjustAndRoundRating -> base10Encode?

File: packages/modules/src/scoring/Scoring.chain.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+      rating: JSON.parse(decoded.rating.toString()),
+      count: JSON.parse(decoded.count.toString()),
+    }
+  } else
+    throw new SDKErrors.ScoreMissingError(
+      `There is not a Score of type ${scoreType} with the provided ID "${entityUri}" on chain.`
+    )
+}
+
+export async function makeScoreEntryToChain(
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.chain.ts]: ratingEntry

File: packages/modules/src/scoring/Scoring.chain.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  ratingInputAndIdentifier: any,
+  authorization: string,
+  authorDid: Cord.DidUri,
+  authorKeys: any,
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.chain.ts]: ?

File: packages/modules/src/scoring/Scoring.chain.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  authorIdentity: any
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.chain.ts]: ?

File: packages/modules/src/scoring/Scoring.chain.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+) {
+  const api = Cord.ConfigService.get('api')
+  const ratingInput = ratingInputAndIdentifier.ratingInput
+  const identifier = ratingInputAndIdentifier.scoreIdentifier
+  const auth = authorization.replace('auth:cord:', '')
```
Comment (smohan-dw) [inline packages/modules/src/scoring/Scoring.chain.ts]: use the SDK method

File: packages/modules/src/scoring/Scoring.ts (hunk 3 above)
Comment (NiranjanAP) [reply packages/modules/src/scoring/Scoring.ts]: @smohan-dw  consider this scenario - count 3, score 15 ( individual entries could have been 6,4 5)
so count * max_value will be 15, it'll pass the validation, but in reality one of the entries '6' is beyond max, so I don't think we justify this validation - right?
I think our validation here will make sense only if we were to get the individual ratings..

Comment (smohan-dw) [reply packages/modules/src/scoring/Scoring.ts]: Correct, managing the correctness of the individual entries is the responsibility of the rating provider. The SDK should make sure the max_value checks are in place. For e.g. an entry with a count of 10 should not have a total value exceeding 50.

**Comment by amarts** [comment]
Looking fine for now. Lets merge and make another work on it. (Merging it on `dev-0.9.0` branch)

**Pull request merged**